Source/WebCore/page/Navigation.cpp

```diff
@@ -147,6 +147,49 @@ void Navigation::updateForActivation(HistoryItem* previousItem, std::optional<Na
     m_activation = NavigationActivation::create(*type, *currentEntry(), WTFMove(previousEntry));
 }
 
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#fire-the-pageswap-event
+RefPtr<NavigationActivation> Navigation::createForPageswapEvent(HistoryItem* newItem, DocumentLoader* documentLoader)
+{
+    auto type = documentLoader->triggeringAction().navigationAPIType();
+    if (!type)
+        return nullptr;
+
+    bool isSameOrigin = SecurityOrigin::create(documentLoader->documentURL())->isSameOriginAs(window()->protectedDocument()->securityOrigin());
+    if (!isSameOrigin || !documentLoader->request().isSameSite())
+        return nullptr;
+
+    // The lookups into m_entries should succeed if the Navigation API is enabled, but
+    // we fallback to a new NavigationHistoryEntry if not so that the pageswap event can
+    // view the to/from URLs even if not.
+    auto getOrCreateNavigationHistoryForHistory = [&](HistoryItem* item) -> RefPtr<NavigationHistoryEntry> {
+        if (!item)
+            return nullptr;
+        if (auto newEntryIndex = item ? getEntryIndexOfHistoryItem(m_entries, *item) : std::nullopt)
+            return m_entries.at(newEntryIndex.value()).ptr();
+        return NavigationHistoryEntry::create(scriptExecutionContext(), *item);
+    };
+
+    RefPtr<NavigationHistoryEntry> oldEntry = getOrCreateNavigationHistoryForHistory(frame() ? frame()->checkedHistory()->currentItem() : nullptr);
```

Comment:
Does `currentEntry()` not return what you want?

Reply:
With the change you requested below, yes, it should.